src/coreclr/nativeaot/Runtime/gcrhenv.cpp

```diff
@@ -1181,8 +1181,12 @@ bool GCToEEInterface::EagerFinalized(Object* obj)
     // Managed code should not be running.
     ASSERT(GCHeapUtilities::GetGCHeap()->IsGCInProgressHelper());
 
+    // the lowermost 3 bits are reserved for storing additional info about the handle
+    // we can use these bits because handle is at least 32bit aligned
+    const uintptr_t HandleTagBits = 7;
```

Comment:
32-bit aligned -> 4 byte aligned -> lowest 2 bits are reusable -> this mask should be 3?

Reply:
Right. I can’t use 3rd bit on 32bit. That is unfortunate.

Reply:
I wonder if the sign bit is available.

Reply:
I think you'll need to adjust Mono as well:
https://github.com/dotnet/runtime/blob/10d342cecc5cbfc57419f8a92af4787f29d76c08/src/mono/mono/metadata/gc-internals.h#L28-L34

Reply:
I think you'll need to adjust Mono here a bit:
https://github.com/dotnet/runtime/blob/10d342cecc5cbfc57419f8a92af4787f29d76c08/src/mono/mono/metadata/gc-internals.h#L28-L34

(shift by two)

Reply:
Mono does not support COM/IWeakReference stuff, so at least for now it may not need anything special. 
That will depend on the final implementation though, as at the moment x86 is a bit of a problem.

Reply:
You mask off the bits in the managed code, so you need to change it. Currently Mono has one reserved bit, you need two.

Reply:
I need 3 bits - 
- one bit to record whether the handle is a long or short. Ideally that info is recorded in the weakreference object, so it is trivially portable.
- one bit to record whether a weakreference has COM stuff associated with it. This is optimization that is important to not do lookups in the side table when there is no COM involved.
- one bit for a lock to prevent concurrent Set. Since we may be setting two values and also setting a bit, we need Set to be exclusive.

Reply:
> You mask off the bits in the managed code, so you need to change it. Currently Mono has one reserved bit, you need two.

Are mono handles 16bit aligned?

Reply:
> > You mask off the bits in the managed code, so you need to change it. Currently Mono has one reserved bit, you need two.
> 
> Are mono handles 16bit aligned?

They are sequential 32-bit indexes (perhaps the limit is lower than 32-bit), not pointers. They get shifted by specified number of bits to allow for the tag bits to be recorded. Currently the macros shift by 1.

Reply:
I see, so effectively 16bit-aligned.

Reply:
Masking is easy to special case for Mono. Thanks for noticing this.

Reply:
It is equally easy to change the shift in the macros to allow more tag bits to match CoreCLR. :-)

Reply:
For the bits shortage problem - One solution is just make the tagged handle a 64bit value, regardelss of the platform and bitness. At least on platforms that support COM. That will have enough spare upper bits for COM stuff. 

It is a nice solution for 64bit, since there is no cost, but on 32bit it is a 25% instance size increase. Maybe that is ok.

Reply:
Why can't we use the same scheme as CoreCLR today? It does fine with just 2 bits.

Reply:
CoreCLR also puts some knowledge of COM into handles. There is a special kind of handle and the handle may be reallocated/changed after creation of WeakReference and potentially concurrently with Get. Since Get does not take a lock, I think there is a subtle handle recycling problem. Also I'd like to avoid calling native code, especially on Get. 

I think we can have a _similar_ solution with only 2 bits on x86, where locking replaces the entire handle to, say `0x1`, and unlocking puts the original value back. It is kinda similar to what CoreCLR does, except for reallocating the handle.
In such scheme Get needs to be aware of locking and spin if it sees `0x1`, which should be rare.

This is a different tradeoff. This way we do not pay extra size on x86, but have some extra complexity in Get.

Reply:
I'd like to have an invariant that the underlying weak handle is an ordinary long/short weak handle, allocated at creation of WeakReference and not changing as long as the WeakReference is alive. 
If that is possible, I think it would make implementation more robust and less dependent on native code (thus easier to port).

Reply:
How about this:
- Bit 0 is used to track whether the handle is "track resurrection" handle (just like today)
- Bit 1 is used to indicate that the handle is "slow path" handle. When it is set, the handle is pointer to large piece of unmanaged memory that contains the handle and the COM knowledge.
- Once the handle is "slow path" handle, we never go back until the weak reference is destroyed.

Reply:
> The eager vs. standard finalization difference is observable in interop

That should not have effect on the referenced objects. We would still not be referencing them strongly, so we cannot impact their lifetime. 

This is more about who closes the weak handle when it is moved into the fat weak reference - GC or finalizer. 
I think we do not care, unless we see lots of fat weak references, and thus regular finalizer is ok.

> The difference is that your scheme has extra GC handle + managed object instead of unmanaged memory that is pretty equivalent. Either one should work fine.

Yes, it is the same idea, the difference is basically to let GC deal with managing memory. I think for this scenario it will do it well enough.